### ISSUE TITLE
Lambda on .NET Core 2.0 with multiple Set-Cookie headers

### ISSUE THREAD

**Issue by raban33** [aws/aws-lambda-dotnet#224]
ASPNET Core 2.0 hosted on Lambda 2.0 Core( Ref: https://forums.aws.amazon.com/thread.jspa?threadID=205782) chokes on multiple headers and joins all the header values with "," separator. This is an issue and browser doesn't transmit those cookies back as the lambda response rendered is below
```
{
  "statusCode": 302,
  "headers": {
    "Content-Security-Policy": "script-src * 'unsafe-inline';style-src * 'unsafe-inline';img-src *",
    "Set-Cookie": "myAuth=896d055430b4f6395a30900eb5883a5b; path=/; secure, myKeys=abc; path=/; secure; httponly",
    "Cache-Control": "no-cache",
    "Pragma": "no-cache",
    "Expires": "-1",
    "Location": "/grants"
  },
  "body": "",
  "isBase64Encoded": false
}

```
In order to have a workaround, What I did was Instead of consuming the nuget package of "Amazon.Lambda.AspNetCoreServer", I had my custom implementation of APIGatewayProxyFunction.cs (Ref: https://github.com/aws/aws-lambda-dotnet/blob/master/Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayProxyFunction.cs) modifying the function  "MarshallResponse".
        
While iterating through the responseFeatures.Headers which contains multiple values in else part, I used  below code 
```
foreach (var kvp in responseFeatures.Headers)
                {
                    if (kvp.Value.Count == 1)
                    {
                        response.Headers[kvp.Key] = kvp.Value[0];
                    }
                    else
                    {
                        //special handling for cookie headers
                        if (kvp.Key.Equals("Set-Cookie", StringComparison.CurrentCultureIgnoreCase))
                        {
                            //string.Concat(newcookieValues.ToArray()).Replace(",", "")
                            for (int i = 0; i < kvp.Value.Count; i++)
                            {
                                if (i == 0)
                                {
                                    response.Headers[kvp.Key] = kvp.Value[0];
                                }
                                else
                                {
                                    var cookieKey = ChangeCookieKeyCasing(response.Headers, "Set-Cookie", i);
                                    if (!response.Headers.ContainsKey(cookieKey))
                                        response.Headers.Add(cookieKey, kvp.Value[i]);
                                }

                            }
                        }
                        else
                            response.Headers[kvp.Key] = string.Join(",", kvp.Value);
                    }

                    // Remember the Content-Type for possible later use
                    if (kvp.Key.Equals("Content-Type", StringComparison.CurrentCultureIgnoreCase))
                        contentType = response.Headers[kvp.Key];
                    // response.Headers = responseFeatures.Headers;
                }
```
the custom function  is: ChangeCookieKeyCasing
```
/// <summary>
        /// Changes the casing of given string at the ordinal position
        /// </summary>
        /// <param name="key"></param>
        /// <param name="ordinalPosition"></param>
        /// <returns></returns>
        public virtual string ChangeCookieKeyCasing(IDictionary<string, string> headers, string key, int ordinalPosition)
        {
            var charArray = key.ToCharArray();
            if (ordinalPosition >= key.Length)
                return key;
            char val = charArray[ordinalPosition];
            bool IsAZ;

            do
            {
                IsAZ = Regex.Matches(val.ToString(), @"[a-zA-Z]").Count > 0;
                if (IsAZ)
                    break;
                ordinalPosition++;
                val = charArray[ordinalPosition];

            } while (!IsAZ);

            //valid string to convert to uppercase
            if (char.IsUpper(val))
                charArray[ordinalPosition] = val.ToString().ToLower()[0];
            else
                charArray[ordinalPosition] = val.ToString().ToUpper()[0];
            var result = new string(charArray);
            if (headers.ContainsKey(result))
                return ChangeCookieKeyCasing(headers, result, ordinalPosition + 1);
            return result;
        }
```
In a nutshell, all I  am doing is if the header key is of type : Set-Cookie, I am adding as many times as the values. Since you have opted to use Dictionary, you can only use one key (case-sensitive) while you can change the case as add the same key to dictionary. Is there a plan to fix this on your end?

**Comment by normj** [comment]
This is a clever work around to get around API Gateway's limitation but I'm not sure I want to add it directly into the package. Mostly I'm concerned of taking on the support of the workaround when I'm not sure how compatible this workaround is for all use cases.

You can do this work around without needing a custom copy of `APIGatewayProxyFunction` by overriding the `PostMarshallResponseFeature` in your class extending `APIGatewayProxyFunction` and then putting your logic into that method. Then you wouldn't have to maintain a separate fork.

**Comment by raban33** [comment]
sorry for the late response but I tried before going down this path and this didn't work and had to resort to the above option

**Comment by rudygt** [comment]
Hi, thanks @raban33 for taking the time of sharing your approach. 

as @normj said, it worked for me to implement it on the PostMarshallResponseFeature 

I will share it here, it may be useful for another poor soul fighting this. 

```csharp
private static readonly string[] SetCookiesCasing = StringPermutationHelper.Combinations(HeaderNames.SetCookie).ToArray();

        protected override void PostMarshallResponseFeature(IHttpResponseFeature aspNetCoreResponseFeature,
            APIGatewayProxyResponse apiGatewayResponse, ILambdaContext lambdaContext)
        {
            base.PostMarshallResponseFeature(aspNetCoreResponseFeature, apiGatewayResponse, lambdaContext);

            if (aspNetCoreResponseFeature.Headers.ContainsKey(HeaderNames.SetCookie))
            {
                var cookieHeaderValue = aspNetCoreResponseFeature.Headers[HeaderNames.SetCookie];

                apiGatewayResponse.Headers.Remove(HeaderNames.SetCookie);

                int i = 0;
                foreach (var value in cookieHeaderValue)
                {                    
                    apiGatewayResponse.Headers.Add(SetCookiesCasing[i], value);
                    i++;
                }
            }
        }

public class StringPermutationHelper
    {
        public static IEnumerable<string> Combinations(string word)
        {
            word = word.ToLower();
            int combinations = 1 << word.Length;
            for (int i = 0; i < combinations; i++)
            {
                char[] result = word.ToCharArray();
                for (int j = 0; j < word.Length; j++)
                {
                    if (((i >> j) & 1) == 1)
                    {
                        result[j] = Char.ToUpper(word[j]);
                    }
                }
                yield return new string(result);
            }
        }
    }
```

**Comment by normj** [comment]
I have changed my mind and got enough feedback that this is worth adding to the package. This behavior will go out with the next release which should be soon. I'll post back when the new version is released.

**Comment by normj** [comment]
Version 2.0.3 of `Amazon.Lambda.AspNetCoreServer` was released with this work around to support multiple cookies. If you had your own work around in place it is probably best to remove it in case having both the built in solution and your work around cause trouble.

**Comment by markskiba** [comment]
Version 2.0.3 doesn't appear to resolve our Lambda/MVC .net Core multiple cookie setting issue.   We upgraded and we're still loosing our Authentication cookie after MVC sets another (e.g. anti-forgery token).  The change note appears to be in the area of "returning"  not MVC "setting" setting of multiple cookies.  A step in the right direction but not a solution to server-less MVC/Lambda working out of the box.
